test/old/testdir/test_autocmd.vim

```diff
@@ -2835,6 +2835,9 @@ func Test_ReadWrite_Autocmds()
   " being modified outside of Vim (noticed on Solaris).
   au FileChangedShell * echo 'caught FileChangedShell'
 
+  " FileWritePre, FileWritePost are only triggered with shelltemp here
+  let shelltemp = &shelltemp
```

Comment:
This is something that we'll need to resolve. Which commands below are triggered  `FileWritePre`/`FileWritePost`, and should they do so for 'noshelltemp' too?

Reply:
I think the tests do too much here, they should only check autocommand which I think could be easier tested then using an external command (gzip) and `!`. It only triggered the file events as a side effect of the bang implementation here.

So my suggestion would be to use regular `write` and `read` commands to test the autocommand functionality and we may create separate tests for the encoding and decoding issues of bang, which is a separate concern which should not break these tests here. 

What do you think?

Reply:
I added an example on how it could look like in the lua counterpart of the similar tests

Reply:
> my suggestion would be to use regular `write` and `read` commands to test the autocommand functionality and we may create separate tests for the encoding and decoding issues of bang,

That could be fine for this test, but there is still the larger question of whether FileWritePre/FileWritePost should work even without 'shelltemp'. Are they useful?

Reply:
You mean if we should trigger those events when we pipe into or from the `!` command? I personally would not, because for me this is not an actual writing/reading a file and I would not expect it to trigger them.